src/graph/immutable_graph.cc

```diff
@@ -0,0 +1,944 @@
+  IdArray indices = IdArray::Empty({static_cast<int64_t>(out_csr->NumEdges())},
+                                   DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+  IdArray eids = IdArray::Empty({static_cast<int64_t>(out_csr->NumEdges())},
+                                DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+  int64_t *indptr_data = static_cast<int64_t*>(indptr->data);
+  dgl_id_t* indices_data = static_cast<dgl_id_t*>(indices->data);
+  dgl_id_t* eid_data = static_cast<dgl_id_t*>(eids->data);
+  std::copy(out_csr->indptr.begin(), out_csr->indptr.end(), indptr_data);
+  std::copy(out_csr->indices.begin(), out_csr->indices.end(), indices_data);
+  std::copy(out_csr->edge_ids.begin(), out_csr->edge_ids.end(), eid_data);
+  return CSRArray{indptr, indices, eids};
+}
+
+////////////////////////////// Graph Sampling ///////////////////////////////
+
+/*
+ * ArrayHeap is used to sample elements from vector
+ */
+class ArrayHeap {
+ public:
+  explicit ArrayHeap(const std::vector<float>& prob) {
+    vec_size_ = prob.size();
+    bit_len_ = ceil(log2(vec_size_));
+    limit_ = 1 << bit_len_;
+    // allocate twice the size
+    heap_.resize(limit_ << 1, 0);
+    // allocate the leaves
+    for (int i = limit_; i < vec_size_+limit_; ++i) {
+      heap_[i] = prob[i-limit_];
+    }
+    // iterate up the tree (this is O(m))
+    for (int i = bit_len_-1; i >= 0; --i) {
+      for (int j = (1 << i); j < (1 << (i + 1)); ++j) {
+        heap_[j] = heap_[j << 1] + heap_[(j << 1) + 1];
+      }
+    }
+  }
+  ~ArrayHeap() {}
+
+  /*
+   * Remove term from index (this costs O(log m) steps)
+   */
+  void Delete(size_t index) {
+    size_t i = index + limit_;
+    float w = heap_[i];
+    for (int j = bit_len_; j >= 0; --j) {
+      heap_[i] -= w;
+      i = i >> 1;
+    }
+  }
+
+  /*
+   * Add value w to index (this costs O(log m) steps)
+   */
+  void Add(size_t index, float w) {
+    size_t i = index + limit_;
+    for (int j = bit_len_; j >= 0; --j) {
+      heap_[i] += w;
+      i = i >> 1;
+    }
+  }
+
+  /*
+   * Sample from arrayHeap
+   */
+  size_t Sample(unsigned int* seed) {
+    float xi = heap_[1] * (rand_r(seed)%100/101.0);
+    int i = 1;
+    while (i < limit_) {
+      i = i << 1;
+      if (xi >= heap_[i]) {
+        xi -= heap_[i];
+        i += 1;
+      }
+    }
+    return i - limit_;
+  }
+
+  /*
+   * Sample a vector by given the size n
+   */
+  void SampleWithoutReplacement(size_t n, std::vector<size_t>* samples, unsigned int* seed) {
+    // sample n elements
+    for (size_t i = 0; i < n; ++i) {
+      samples->at(i) = this->Sample(seed);
+      this->Delete(samples->at(i));
+    }
+  }
+
+ private:
+  int vec_size_;  // sample size
+  int bit_len_;   // bit size
+  int limit_;
+  std::vector<float> heap_;
+};
+
+static void RandomSample(size_t set_size,
+                         size_t num,
+                         std::vector<size_t>* out,
```

Comment:
Any particular reason why a pointer instead of a reference is used here (and below)?